README.md
# 7pass-ios-sdk

7Pass iOS SDK is a Swift library for interacting with the
[7Pass SSO service](https://7pass.de). You can use this library to
implement authentication for your app and take advantage of the
already existing features that 7Pass SSO offers.

## Usage

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Installation

SevenPassSDK is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
source 'https://github.com/CocoaPods/Specs.git'

use_frameworks!

target 'MyApp' do
  pod 'SevenPassSDK'
end
```

## Running the sample application

To demonstrate the functionality of the library, there's a sample
application available. The application is configured to run against
the QA instance of 7Pass and uses test credentials
(Example/SevenPassSDK/SsoManager.swift). Feel free to use these credentials
while testing but you need to obtain real credentials before releasing
your app.

To obtain the real credentials, you first need to contact the 7Pass
Tech Team.

Once you have the credentials available, you can go ahead and type

```
pod install
```

in `Example` directory, open `SevenPassSDK.xcworkspace` project in XCode,
select `SevenPassSDK-Example` build configuration and run it.

The sample application should now start within the configured device
(or the emulator) and should provide several tabs each implementing a
feature you might want to use in your app.

## Usage

You are strongly encouraged to go over the sample application to see
how the API should be used.

If you're starting the development, it's always a good idea to work
against a non live instance of the 7Pass SSO service. In this case,
we'll use the QA environment. Don't forget to switch to the production
one before you release your application to the public.

First, let's initialize the library with sample configuration:

```swift
let configuration = SevenPassConfiguration(
    consumerKey: "56a0982fcd8fb606d000b233",
    consumerSecret: "2e7b77f99be28d80a60e9a2d2c664835ef2e02c05bf929f60450c87c15a59992",
    callbackUri: "oauthtest://oauth-callback",
    environment: "qa"
)

let sso = SevenPass(configuration: configuration)
```

## Authentication flow

The library offers several ways of logging in the user. The result of
every login is eventually an instance of `TokenSet`. A `TokenSet`
represents the three different token types the server will return:

1. *Access token* - This token proofs the identity of the user and is
thus included in almost every remote call the library
performs. Every access token is valid just for 2 hours. In cannot
be used after that.

2. *Refresh token* - This token can be used to get a new `TokenSet`
with a fresh access token. Its expiration time is set to 90 days
and its prolonged every time you use it. If however the token
expired, you need to ask the user to login again.

3. *ID token* - This token contains information (claims) about the
logged in user. You can for example get the user's ID, name, email
etc.

You don't generally need to worry about the current `TokenSet` and how
it's used but you need to make sure it's valid before use:

```swift
sso.authorize(
    scopes: ["openid", "profile", "email"],
    success: { tokenSet in
        // tokenSet.accessToken?.isExpired()
        // tokenSet.refreshToken?.isExpired()
    },
    failure: { error in
        print(error.localizedDescription)
    }
)
```

If that access token is fresh, you can use the token set to perform API
calls. After that, you can use that TokenSet to initialize an API client:

```swift
let accountClient = sso.accountClient(tokenSet)
```

Otherwise, you need to refresh it, using a refresh token.

```swift
let refreshTokenString = tokenSet.refreshToken!.token

sso.authorize(refreshToken: refreshTokenString,
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

Of course, you need to make sure that the refresh token itself is
fresh. If it's not, you cannot get a fresh access token and need to
ask the user to log in again. If you're using the Web view flow and
the user has an active session on 7Pass, chances are he/she will come
back immediately without having to fill in the credentials.

Since the same token set can be used in the span of multiple days, you
should store it instead of forcing the user to re-login every
time. There's a utility class you can use to save the token set
into the keychain.

```swift
let tokenSetCache = SevenPassTokenSetCache(configuration: sso.configuration)

// Load token set from a keychain
let tokenSet = tokenSetCache.load()

// Store token set in the keychain
tokenSetCache.tokenSet = tokenSet
tokenSetCache.save()

// Delete token set
tokenSetCache.delete()
```

### 1. Logging in using Web view

The most basic way of logging in the user is using the Web view.

```swift
sso.authorize(
    scopes: ["openid", "profile", "email"],
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

The code will open the provided Web view which will navigate to the
7Pass's login dialog. From there, the user has several options, he/she
can use his/her credentials directly or use Google/Facebook.

Once the process is done, the web view will automatically close and
the result will be provided in the callback.

Once the process is successfully finished, we can use the obtained
`tokenSet` to instantiate API client.

### 2. Logging in using login and password

In case you want to provide more native experience for the user, the
library offers logging in using the user's 7Pass login and
password (grant_type = password). This way, you get to design the login
form yourself using the iOS widgets.

```swift
sso.authorize(
    login: login.text!,
    password: password.text!,
    scopes: ["openid", "profile", "email"],
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

### 3. Logging in using social

In situations when the user is already logged in using some other
service (currently supported services are Facebook or Google), you can
use the `social` flow.

In this flow, you provide an access token you've received from the other service
and 7Pass will make sure to create a new 7Pass account (with all of the user
details) or identify an existing 7Pass account.

```swift
sso.authorize(
    providerName: "facebook",
    accessToken: "foobarbaz",
    scopes: ["openid", "profile", "email"],
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

### 4. Autologin

In case, that you already have valid `TokenSet` (fe. from password or social login) or a valid `autologin_token` and you want to create user session in the WebView, you can utilize `autologin` method.

By default this method sets `response_type = "none"`

### 5. Logging out

In the app's perspective, a user is logged in when you have a fresh
`TokenSet`. In order to log the user out, all that's required is to
"forget" the tokens. Optionally, you can destroy the user's session in WebView as well.

To use it, just call

```swift
sso.destroyWebviewSession(failure: errorHandler)
```

### Handling interaction_required error

It's important to note that if the user hasn't accepted all of the necessary consents for the client or the service the client belongs to, the login through Username & Password or Social method might fail and it's necessary to handle this situation in your error handling callback.

```swift
func errorHandler(error: NSError) {
    // Let autologin handle interaction_required errors
    if let errorMessage = error.userInfo["error"] as? String where errorMessage == "interaction_required" {
        let autologinToken = error.userInfo["autologin_token"] as! String

        self.sso.autologin(
            autologinToken: autologinToken,
            scopes: ["openid", "profile", "email"],
            params: ["response_type": "id_token+token"],
            success: { tokenSet in
                // Do some stuff
            },
            failure: errorHandler
        )
    }

    // Handle other errors
}
```

##### Login using valid tokenSet

TokenSet has to include valid `access_token` and `id_token`

```swift
sso.autologin(tokenSet,
    scopes: ["openid", "profile", "email"],
    rememberMe: false,
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

##### Login using autologin token

Used for example when `interaction_required` error is returned from the server.

```swift
sso.autologin(
    autologinToken: "AUTOLOGINTOKEN",
    scopes: ["openid", "profile", "email"],
    params: ["response_type": "id_token+token"],
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

## Account client

Now that you have an account's `TokenSet`, you can already identify the user and
proceed to your actual app. You can, however, request further details about the
user. Each method call corresponds to an endpoint available on 7Pass, you're
encouraged to go through the
[documentation](http://guide.docs.7pass.ctf.prosiebensat1.com/api/index.html?focusTabs=node#api-Accounts).

```swift
let accountClient = sso.accountClient(tokenSet)
```

### Getting the account's details

The simplest remote call you can make is to request [the account's
details](http://guide.docs.7pass.ctf.prosiebensat1.com/api/index.html?focusTabs=node#api-Accounts-GetAccount).

```swift
accountClient.get("me",
    success: { json, response in
        print(json)
    },
    failure: errorHandler
)
```

Note that you might not need to request the additional user details as the most
basic ones are already present in the ID Token as part of the `TokenSet`. It
generally depends on your app's requirements.

```swift
tokenSet.idTokenDecoded
```

### Refreshing accountClient TokenSet

Account client is an instance of `SevenPassRefreshingClient` which handles refreshing of the tokens for you seamlessly.

Optionally callback can be supplied as an argument to `accountClient` method:

```swift
let accountClient = sso.accountClient(
  tokenSet,
  tokenSetUpdated: { tokenSet in
    // Save new tokens
    tokenSetCache.tokenSet = tokenSet
    tokenSetCache.save()
  }
)
```

## Credentials client

A credentials client can be used to perform various administrative tasks not
bound to a particular account. For example, you can verify whether an email
address is unused or whether a password is of sufficient quality and others.
Getting this kind of client doesn't require any interaction from the user, it's
issued based on your client's credentials.

Same as with the Account client you will receive a valid `TokenSet`, however,
the token set will not contain anything but an access token and its validity
will be limited to just 15 minutes. Since the token set doesn't have a Refresh
token, it's necessary to request a new one after the 15 minutes have passed if
required.

```swift
sso.authorize(
    parameters: [
        "grant_type": "client_credentials",
    ],
    success: { tokenSet in
        // Do some stuff
    },
    failure: errorHandler
)
```

Once you have the token set available, you can get the Credentials client and
invoke the available methods.

```swift
let credentialsClient = sso.credentialsClient(tokenSet)
```

### Checking an email's availability

Validates the provided email address and returns whether it's available for use.
You can use this method to give immediate feedback to users (i.e. during a
registration process). See more information in [the
documentation](http://guide.docs.7pass.ctf.prosiebensat1.com/api/#api-Emails-ActionCheckMail).

```swift
deviceClient.post("checkMail",
    parameters: [
        "email": login.text!,
        "flags": [
            "client_id": sso.configuration.consumerKey
        ]
    ],
    success: { json, response in
        if let error = json["data"]?["error"] as? String {
           // E-Mail is invalid
        }
    },
    failure: errorHandler
)
```

The response also contains a suggested email address so that you can easily
provide the "did you mean x?" functionality in case the user has made a typo in
the address.

### Checking a password's validity

As a next step in a registration form you might want to provide a feedback
regarding the validity of the provided password. You can find all of the
requirements and the response codes in [the
documentation](http://guide.docs.7pass.ctf.prosiebensat1.com/api/#api-Accounts-ActionCheckPassword).

```swift
deviceClient.post("checkPassword",
    parameters: [
        "password": password.text!
    ],
    success: { json, response in
        if let errors = json["data"]?["errors"] as? [String] {
            // Password is not valid, got array of errors
        }
    },
    failure: errorHandler
)
```

### Creating a new account

Finally, you can create a brand you account directly. The only mandatory
parameter is a valid email address and a password. If you don't want to force
your users to type in the password, you can request an auto generated password
(in which case only the email address is required).

```swift
deviceClient.post("registration",
    parameters: [
        "email": login.text!,
        "password": password.text!,
        "flags": [
            "client": [
                "id": sso.configuration.consumerKey, // Associate with a service
                "agb": true, // Accept ToS consents (user should given permissions beforehand)
                "dsb": true, // Accept privacy policy consents (user should given permissions beforehand)
                "implicit": true // Accept implicit optins
            ]
        ]
    ],
    success: { json, response in
        // Account creation status returned in json constant
    },
    failure: errorHandler
)
```

## Advanced usage

### Customize WebView

Create your own custom class implemeting `SevenPassURLHandlerType` protocol and pass that instance to `SevenPass` init.

```swift
let sso = SevenPass(configuration: configuration, urlHandler: YourWebViewController())
```

## Developers

### OAuthSwift custom fork

This library comes bundled with custom fork of OAuthSwift library, that can be downloaded and builded using
[Carthage](https://github.com/Carthage/Carthage), to rebuild it, run:

```swift
carthage update --platform iOS
```

## License

SevenPassSDK is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# CHANGELOG

The 7Pass iOS SDK follows [SemVer](http://semver.org/).

- 2.0.0 (2016-11-07)
  - Compatibility updates for Swift 3.0.1

- 1.1.1 (2016-04-25)
  - Compatibility updates for Xcode 7.3
  - Improve SevenPassClient error handling

- 1.1.0 (2016-03-17)
  - Sends consumerKey as X-Service-Id header value

- 1.0.3 (2016-02-24)
  - Remove WebView cookies recursively for each of the parent domains
  - Change `development` env host to `https://7pass.dev`

- 1.0.2 (2016-02-23)
  - Fix OAuthSwift import paths

- 1.0.1 (2016-02-17)
  - Bundle custom OAuthSwift.framework

- 1.0.0 (2016-02-06)
  - Rename method `deviceCredentialsClient` to `credentialsClient`
  - Documentation fixes

- 0.0.1 (2016-02-02)
  - Initial release.


Pod/Classes/SevenPass.swift
//
//  SevenPass.swift
//  SevenPass
//
//  Created by Jan Votava on 15/12/15.
//  Copyright © 2015 Jan Votava. All rights reserved.
//

import OAuthSwift
import JWTDecode
import JWT

public let OAuthJWTErrorDomain = "OAuthJWTErrorDomain"

public class SevenPass: NSObject {
    public let configuration: SevenPassConfiguration
    public let urlHandler: SevenPassURLHandlerType

    var oauthSwiftCache: OAuth2Swift?

    public init(configuration: SevenPassConfiguration, urlHandler: SevenPassWebViewController? = nil) {
        self.configuration = configuration
        self.urlHandler = urlHandler ?? SevenPassWebViewController(urlString: configuration.callbackUri)
    }

    private func loadJwt(_ jwt: String, consumerKey: String? = nil) throws -> JWT {
        let jwt = try decode(jwt: jwt)

        guard !jwt.expired else {
            let error = NSError(domain:OAuthJWTErrorDomain, code:0, userInfo:[NSLocalizedDescriptionKey: "Token is expired"])

            throw error
        }

        if let consumerKey = consumerKey {
            guard jwt.audience!.contains(consumerKey) else {
                let error = NSError(domain:OAuthJWTErrorDomain, code:0, userInfo:[NSLocalizedDescriptionKey: "Token has invalid audience"])

                throw error
            }
        }

        return jwt
    }

    public class func handleOpenURL(_ url: URL) {
        return OAuthSwift.handle(url: url)
    }

    func baseUri(_ basePath: String) -> URL {
        return URL(string: "\(configuration.host)\(basePath)")!
    }

    public func accountClient(_ tokenSet: SevenPassTokenSet, tokenSetUpdated: SevenPassRefreshingClient.TokenSetUpdated? = nil) -> SevenPassRefreshingClient {
        return SevenPassRefreshingClient(sso: self, baseUri: baseUri("/api/accounts/"), tokenSet: tokenSet, consumerKey: configuration.consumerKey, consumerSecret: configuration.consumerSecret, tokenSetUpdated: tokenSetUpdated)
    }

    public func credentialsClient(_ tokenSet: SevenPassTokenSet) -> SevenPassClient {
        guard let accessToken = tokenSet.accessToken?.token else { fatalError("accessToken is missing") }

        return SevenPassClient(baseUri: baseUri("/api/client/"), accessToken: accessToken, consumerKey: configuration.consumerKey, consumerSecret: configuration.consumerSecret)
    }

    func oauthSwift() -> OAuth2Swift {
        if let cached = self.oauthSwiftCache {
            return cached
        }

        let config = self.configuration.config

        let oauthSwift = OAuth2Swift(
            consumerKey:    self.configuration.consumerKey,
            consumerSecret: self.configuration.consumerSecret,
            authorizeUrl:   config["authorization_endpoint"] as! String,
            accessTokenUrl: config["token_endpoint"] as! String,
            responseType: ""
        )

        oauthSwift.authorizeURLHandler = self.urlHandler
        self.oauthSwiftCache = oauthSwift

        return oauthSwift
    }

    func loadResponse(_ parameters: Dictionary<String, Any>, login: String = "") throws -> SevenPassTokenSet {
        let tokenSet = SevenPassTokenSet()

        if let accessToken = parameters["access_token"] as? String {
            var expiresIn: TimeInterval = 120 // Default expire to 2 minutes

            // Implicit flow
            if let expires = parameters["expires_in"] as? String { // Implicit flow
                expiresIn = TimeInterval(expires)! - TimeInterval(60)
            } else if let expires = parameters["expires_in"] as? Int { // Authorization code flow
                expiresIn = TimeInterval(expires) - TimeInterval(60)
            }

            tokenSet.accessToken = SevenPassToken(token: accessToken, expiresIn: expiresIn)
        }

        if let refreshToken = parameters["refresh_token"] as? String {
            let expiresIn = TimeInterval(90 * 24 * 60 * 60) - TimeInterval(60) // 90 days - 60s

            tokenSet.refreshToken = SevenPassToken(token: refreshToken, expiresIn: expiresIn)
        }

        if let idToken = parameters["id_token"] as? String {
            tokenSet.idToken = idToken
            let jwt = try self.loadJwt(idToken, consumerKey: self.configuration.consumerKey)
            tokenSet.idTokenDecoded = jwt.body
        }

        return tokenSet
    }

    public func authorize(parameters: Dictionary<String, String>, success: @escaping (SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {
        var parameters = parameters
        configuration.fetch(
            success: { [unowned self] in
                let config = self.configuration.config

                parameters["client_id"] = self.configuration.consumerKey
                parameters["client_secret"] = self.configuration.consumerSecret

                let successHandler: OAuthSwiftHTTPRequest.SuccessHandler = { data, response in
                    let json: Dictionary<String, AnyObject>

                    do {
                        json = try JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.mutableContainers) as! Dictionary<String, AnyObject>
                    } catch let error {
                        failure(error as NSError)

                        return
                    }

                    let tokenSet: SevenPassTokenSet

                    do {
                        tokenSet = try self.loadResponse(json)
                    } catch let error {
                        failure(error as NSError)

                        return
                    }

                    success(tokenSet)
                }

                self.oauthSwift().client.request(config["token_endpoint"] as! String,
                    method: .POST,
                    parameters: parameters,
                    headers: ["X-Service-Id": self.configuration.consumerKey],
                    success: successHandler,
                    failure: SevenPassError.handler(success: success, failure: failure)
                )
            },
            failure: failure
        )
    }

    public func authorize(scopes: Array<String>, params: OAuthSwift.Parameters = [:], success: @escaping (_ tokenSet: SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {
        var params = params
        configuration.fetch(
            success: { [unowned self] in
                params["nonce"] = OAuthSwiftCredential.generateNonce()

                if params["response_type"] == nil {
                    var responseType = "code"

                    // Implicit flow when secret is not present
                    if self.configuration.consumerSecret.isEmpty {
                        responseType = "id_token+token"
                    }

                    params["response_type"] = responseType
                }

                self.oauthSwift().authorize(
                    withCallbackURL: URL(string: self.configuration.callbackUri)!,
                    scope: scopes.joined(separator: "+"),
                    state: "ios",
                    parameters: params,
                    success: { credential, response, parameters in
                        let tokenSet: SevenPassTokenSet

                        do {
                            tokenSet = try self.loadResponse(parameters)
                        } catch let error {
                            failure(error as NSError)

                            return
                        }

                        success(tokenSet)
                    },
                    failure: SevenPassError.handler(success: success, failure: failure)
                )
            },
            failure: failure
        )
    }

    public func authorize(login: String, password: String, scopes: Array<String>, success: @escaping (_ tokenSet: SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {
        authorize(
            parameters: [
                "grant_type": "password",
                "scope": scopes.joined(separator: " "),

                "login": login,
                "password": password
            ],
            success: success,
            failure: failure
        )
    }

    public func authorize(refreshToken: String, success: @escaping (_ tokenSet: SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {
        authorize(
            parameters: [
                "grant_type": "refresh_token",

                "refresh_token": refreshToken
            ],
            success: success,
            failure: failure
        )
    }

    public func authorize(providerName: String, accessToken: String, scopes: Array<String>, success: @escaping (_ tokenSet: SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {
        authorize(
            parameters: [
                "grant_type": "social",
                "scope": scopes.joined(separator: " "),

                "provider_name": providerName,
                "access_token": accessToken
            ],
            success: success,
            failure: failure
        )
    }

    public func autologin(autologinToken: String, scopes: Array<String>, params: [String: String] = [String: String](), success: @escaping (_ tokenSet: SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {

        var params = params
        params["autologin"] = autologinToken

        if params["response_type"] == nil {
            params["response_type"] = "none"
        }

        authorize(
            scopes: scopes,
            params: params,
            success: success,
            failure: failure
        )
    }

    public func autologin(_ tokenSet: SevenPassTokenSet, scopes: Array<String>, rememberMe: Bool, params: [String: String] = [String: String](), success: @escaping (_ tokenSet: SevenPassTokenSet) -> Void, failure: @escaping SevenPassError.Handler) {
        guard let accessToken = tokenSet.accessToken?.token else { fatalError("accessToken is missing") }
        guard let idToken = tokenSet.idToken else { fatalError("idToken is missing") }

        let payload: Dictionary<String, Any> = [
            "access_token": accessToken,
            "id_token": idToken,
            "remember_me": rememberMe
        ]

        let autologinToken = encode(payload, algorithm: .hs256(configuration.consumerSecret.data(using: .utf8)!))

        autologin(autologinToken: autologinToken, scopes: scopes, params: params, success: success, failure: failure)
    }

    public func destroyWebviewSession(success: (() -> Void)? = nil, failure: @escaping SevenPassError.Handler) {
        configuration.fetch(
            success: {
                let config = self.configuration.config
                let endSessionEndpoint = config["end_session_endpoint"] as! String

                self.urlHandler.destroySession(logoutUrl: URL(string: endSessionEndpoint)!)
            },
            failure: failure
        )
    }
}
